required-features targets are compiled by cargo test even when such features are not selected

I started dogfooding required-features and I noticed that `cargo test` is compiling targets that require features which are not selected. This is particularly a problem when a bin using require-features has an optional dependency because the optional dependency won't be found and the test will fail to compile.

**To reproduce:**
1. Create a new project with the following Cargo.toml.
```toml
[project]
name = "foo"
version = "0.0.1"
authors = []

[features]
a = []

[[bin]]
name = "foo"
path = "src/bin/foo.rs"
required-features = ["a"]
```
2. Add a src/bin/foo.rs file with the following contents.
```rust
extern crate bar;

fn main() {
}
```
3. Add an empty tests/foo.rs file.
4. Run `cargo test`

**The error is:**
```shell
   Compiling foo v0.0.1 (...)
error[E0463]: can't find crate for `bar`
  --> src/bin/foo.rs:1:1
   |
1  | extern crate bar;
   | ^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error

error: Could not compile `foo`.

error: build failed
```

**Expected outcome:**
The binary should not be compiled with the tests unless the appropriate features are selected.

**Work around:**
Enable the required features when testing: `cargo test --features="a"`

I'm working on a pull request for this issue but it'll be a couple days before I can submit it.

--- a/src/cargo/ops/cargo_rustc/context.rs
+++ b/src/cargo/ops/cargo_rustc/context.rs
@@ -639,16 +639,13 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         // Integration tests/benchmarks require binaries to be built
         if unit.profile.test &&
            (unit.target.is_test() || unit.target.is_bench()) {
-            let no_features = HashSet::new();
-            let features = self.resolve.features(id).unwrap_or(&no_features);
-
             ret.extend(unit.pkg.targets().iter().filter(|t| {
                 let no_required_features = Vec::new();
 
                 t.is_bin() &&
                 // Skip binaries with required features that have not been selected.
                 t.required_features().unwrap_or(&no_required_features).iter().all(|f| {
-                    features.contains(f)
+                    self.resolve.features(id).contains(f)
                 })
             }).map(|t| {
                 Unit {

--- a/src/cargo/ops/cargo_rustc/context.rs
+++ b/src/cargo/ops/cargo_rustc/context.rs
@@ -639,7 +639,18 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         // Integration tests/benchmarks require binaries to be built
         if unit.profile.test &&
            (unit.target.is_test() || unit.target.is_bench()) {
-            ret.extend(unit.pkg.targets().iter().filter(|t| t.is_bin()).map(|t| {
+            let no_features = HashSet::new();
+            let features = self.resolve.features(id).unwrap_or(&no_features);
+
+            ret.extend(unit.pkg.targets().iter().filter(|t| {
+                let no_required_features = Vec::new();
+
+                t.is_bin() &&
+                // Skip binaries with required features that have not been selected.
+                t.required_features().unwrap_or(&no_required_features).iter().all(|f| {
+                    features.contains(f)
+                })
+            }).map(|t| {
                 Unit {
                     pkg: unit.pkg,
                     target: t,

--- a/tests/required-features.rs
+++ b/tests/required-features.rs
@@ -979,3 +979,62 @@ test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured
     assert_that(p.cargo("uninstall").arg("foo"),
                 execs().with_status(0));
 }
+
+#[test]
+fn test_skips_compiling_bin_with_missing_required_features() {
+    let p = project("foo")
+        .file("Cargo.toml", r#"
+            [project]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+
+            [features]
+            a = []
+
+            [[bin]]
+            name = "bin_foo"
+            path = "src/bin/foo.rs"
+            required-features = ["a"]
+        "#)
+        .file("src/bin/foo.rs", "extern crate bar; fn main() {}")
+        .file("tests/foo.rs", "")
+        .file("benches/foo.rs", "");
+    p.build();
+
+    assert_that(p.cargo("test"),
+                execs().with_status(0).with_stderr(format!("\
+[COMPILING] foo v0.0.1 ({})
+[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] target[/]debug[/]deps[/]foo-[..][EXE]", p.url()))
+                .with_stdout("
+running 0 tests
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured
+
+"));
+
+    assert_that(p.cargo("test").arg("--features").arg("a").arg("-j").arg("1"),
+                execs().with_status(101).with_stderr_contains(format!("\
+[COMPILING] foo v0.0.1 ({})
+error[E0463]: can't find crate for `bar`", p.url())));
+
+    if is_nightly() {
+        assert_that(p.cargo("bench"),
+                    execs().with_status(0).with_stderr(format!("\
+[COMPILING] foo v0.0.1 ({})
+[FINISHED] release [optimized] target(s) in [..]
+[RUNNING] target[/]release[/]deps[/]foo-[..][EXE]", p.url()))
+                    .with_stdout("
+running 0 tests
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured
+
+"));
+
+        assert_that(p.cargo("bench").arg("--features").arg("a").arg("-j").arg("1"),
+                    execs().with_status(101).with_stderr_contains(format!("\
+[COMPILING] foo v0.0.1 ({})
+error[E0463]: can't find crate for `bar`", p.url())));
+    }
+}

